### ISSUE TITLE
Measure whether NPU and GPU can share one interpreter on the MT6899, and what a crossing costs

### ISSUE THREAD

**Issue by m96-chan** [m96-chan/NxPU#174]
Filed from DroidRunner, where [#159](https://github.com/m96-chan/DroidRunner/issues/159) was investigated. Three measurements remain, all of them need the MT6899, and that handset is here rather than there.

## What is already settled, so nobody re-derives it

Two delegates on one interpreter is a supported TFLite shape — checked against the 2.16.1 AAR DroidRunner ships, where `Interpreter.Options.addDelegate` accumulates and `NativeInterpreterWrapper` holds a `List<Delegate>`.

But the answer differs by vendor, and on Qualcomm it is permanently no:

| phone | NPU reached by | same process | NPU + GPU in one interpreter |
| --- | --- | --- | --- |
| SM8650 | `qnn-htp` | **no** | **impossible by construction** |
| **MT6899** | `mtk-*`, through NNAPI | yes | the open question |
| Tensor G4 | `google-edgetpu`, through NNAPI | yes | possible in principle |

On the SM8650 the NPU is reachable only through Qualcomm's runtime, which lives in a separate process for a licensing reason (the FSF's line for "one program" is the shared address space, and PRoot's GPL-2.0 code is in the main process). Two delegates need one address space, so that is not an engineering trade-off anyone can revisit.

**Consequence worth carrying into the routing model:** the cost of an engine boundary is not a device-independent constant. On MediaTek and Tensor it is a round trip through TFLite's CPU-visible tensors — a memory cost. On Qualcomm it would be IPC, and no interpreter spans both. A schedule tuned on an MT6899 does not transfer to an SM8650 even in shape.

## It is blocked on a DroidRunner change, not on the phone

`--device` names one accelerator. Expressing a partitioned run needs an ordered list, and the order decides who claims what, so it cannot be normalised away. Until DroidRunner can express that, none of the three below can be asked for through the agent.

So the split is:

- **DroidRunner** ships a way to name two delegates in order. Not yet agreed — it is a contract change and the argument for it is exactly the measurement below, which makes this slightly circular. A debug-only or experimental form would be enough to break the circle.
- **NxPU** runs the sweep, because the phone is there. It is attached over USB to the DroidRunner machine, so the app can be put on it from here; dispatching to it goes through this repository.

## The three measurements

**1. Do the two delegates actually coexist?**

Attach the NNAPI delegate pinned to `mtk-neuron_shim`, then the GPU delegate, to one interpreter. TFLite prints one line per delegate that claims anything:

```
Replacing %d out of %d node(s) with delegate (%s) node, yielding %zu partitions
```

Two such lines means they coexist and the split is in them. One line, or a failure to build the interpreter, means they do not — and that is a complete answer to this issue on its own. Delegates are allowed to decline to share.

**2. The split each one claims.**

From the same log. Worth having because the expected semantics — each delegate offered what the previous did not claim, so order is load-bearing — is TFLite's documented behaviour and not something either of us has seen on this silicon.

**3. The crossing cost**, by the comparison already written into #159:

| the same two-operator graph | |
| --- | --- |
| both nodes on the NPU | baseline |
| both nodes on the GPU | baseline |
| node 1 NPU, node 2 GPU | baseline + one crossing |

Use an operator both engines accept — the sweep on this device found 15 of those, so there is a choice. `iterations: 0` is not appropriate here: this one is about time, and it is the only one of the three that is.

## What each answer changes

- **1 says no** → a partitioning compiler emits one model per engine and the host chains them. That is a real design constraint and worth knowing early.
- **1 says yes and 3 is small** → per-operator routing across engines is worth doing, and the GPU accepting a superset of the NPU here makes it a pure speed question.
- **1 says yes and 3 dominates** → the fastest schedule is "put everything on one engine", which is a good answer nobody can currently demonstrate.

**Comment by m96-chan** [comment]
Measurement 3's two baselines are done. **They change what the other two are for.**

MT6899, 30 iterations, the 12 operators both engines accept, `conditions.stable`
true on every model — so none of this is a throttle
([run](https://github.com/m96-chan/NxPU/actions/runs/34047089054), branch
`ci/latency-sweep`).

```
model              NPU med   NPU p90   GPU med   GPU p90    faster
concat                1.38      1.55      1.17      1.77       GPU
concat_axis1          1.46      1.54      2.25      2.72       NPU
gather                0.52      0.61      0.93      1.10       NPU
maxpool              33.54     35.13     39.96     42.03       NPU
relu                  1.28      1.39      1.21      2.12       GPU
split                 1.34      1.38      0.83      1.55       GPU
split_axis1           1.38      1.44      1.96      2.47       NPU
tanh_act              1.27      1.34      1.10      1.34       GPU
vecadd                1.33      1.40      1.11      1.52       GPU
vecadd_const          1.32      1.38      1.07      1.80       GPU
vecmul                1.38      1.70      1.14      2.09       GPU
vecsub                1.49      2.90      1.31      1.76       GPU
```

**The GPU is faster in the middle on 8 of 12.** The NPU wins only where there
is real work — `maxpool`, the axis-crossing `concat`/`split`, `gather`.

**The tail says the opposite.**

| | NPU | GPU |
| --- | --- | --- |
| p90 / median | 1.06× median, 1.95× worst | **1.36×** median, 1.87× worst |
| p99 / median | 1.27× median, 2.69× worst | **2.04×** median, **6.23×** worst |

With `stable` true throughout, so the remaining explanation is contention — and
it has a cause: the GPU is the engine the display composites on. The NPU is a
dedicated part.

## What this does to the routing model

The premise underneath measurement 3 was that a crossing might be worth paying
because one engine is faster. On this phone the faster engine is usually the
GPU, and it is the one you cannot occupy freely. So:

- **A schedule minimising the sum of medians picks the GPU every time**, and is
  wrong for anything with a frame deadline.
- **"Send everything the NPU will take" is not a fallback**, it is a defensible
  policy — even where the GPU is faster per operator. That makes the NPU's
  narrower operator set the thing to widen, which on this silicon means
  supplying constant filters (m96-chan/NxPU#176).
- Measurement 1 is still worth having, and measurement 3's third cell too. But
  if they come back saying crossings are cheap, that is **not** an argument for
  per-operator routing by speed. The quantity to minimise is the tail and the
  time the GPU is held, not the sum of medians.

## Still blocked, and I did not work around it

`--device` names one accelerator; there is no list form in the CLI or the
contract. So measurement 1 and the split cell of measurement 3 cannot be asked
for, exactly as this issue says.

The fixtures for the split cell exist and need no new work when it can be:
`batchnorm` lowers to MUL + ADD in one model, and its halves are `vecmul` and
`vecadd` as separate ones. All three are in `examples/`, and all three
operators are accepted by both engines.

## One thing found on the way

A self-hosted runner keeps its workspace between jobs, so this run's artifact
carried four stale `sweep-*.json` from the previous sweep — for drivers it
never asked about. `reduce.py` reads `drivers.txt` and was unaffected, but
anyone opening the artifact would have seen answers to a question this run did
not ask. The sweep step now clears them first.

**Comment by m96-chan** [comment]
**All three answered.** MT6899, 30 iterations, `conditions.stable` true on every
cell ([run](https://github.com/m96-chan/NxPU/actions/runs/34086976474), branch
`ci/delegate-split`).

## 1. Do the two delegates coexist? Yes

`--feature multi-delegate` is accepted, `--device 'mtk-neuron_shim+gpu'` builds
an interpreter, and `delegations` comes back as an array. Attaching an idle
second delegate is free: `batchnorm` is 76.91ms on the NPU alone and 74.25ms
with the GPU attached and claiming nothing, `vecadd` 1.34ms against 1.37ms —
noise in both directions.

## 2. The split each one claims

It took three fixtures to get one, and the two that failed are worth as much as
the one that worked.

**`vecadd`, `batchnorm` — no split, and that is the mechanism.** The NNAPI
delegate claimed every node (`delegated: 2 of 2, partitions: 1`) and the GPU
was offered nothing. `a+b` sets the order, not the assignment, so a model whose
operators the first delegate takes entirely leaves the second idle. Same result
you saw on the SM8650, for the same reason.

**`attention` — splits, then cannot run.** On the NPU alone it comes back
`partial`: *2 of 4 operators on TfLiteNnapiDelegate, 2 on the CPU, 3
partitions* — the DIV and the SOFTMAX taken, both float32 BATCH_MATMULs
refused. With the GPU attached the remainder is offered to it and it claims
one. Then:

```
TfLiteGpuDelegate Invoke: Failed to upload data to GPU
(clEnqueueWriteBuffer) - Invalid command queue
```

**Not the pairing** — the GPU delegate fails identically on this model with
nothing else attached. A partition that cannot run is not a boundary that is
expensive.

**`conv2d_5x5` — splits and runs.** It lowers to PAD + CONV_2D, and with the
filter still a graph input the NNAPI drivers refuse the convolution while the
GPU takes one, so the two engines want different halves:

```json
[{"delegate": "TfLiteNnapiDelegate",  "delegated": 1, "total": 2, "partitions": 2},
 {"delegate": "TfLiteGpuDelegateV2",  "delegated": 1, "total": 2, "partitions": 2}]
```

## 3. What a crossing costs

| `conv2d_5x5`, PAD + CONV_2D | median | p90 | p99 | min |
| --- | --- | --- | --- | --- |
| both on the GPU | **60.04ms** | 66.09 | 70.25 | 47.04 |
| NPU alone (`partial`, 1 on the CPU) | 424.57ms | 439.50 | 447.40 | 418.83 |
| **NPU + GPU, one operator each** | **100.48ms** | 101.42 | 104.01 | 96.02 |

**The split is 1.67× slower than putting both operators on the GPU.**

That 40ms is an upper bound on the crossing, not the crossing: it also contains
whatever NNAPI's PAD costs against the GPU's. But the direction is not in
doubt, and it is the number a scheduler would actually pay.

Worth noting the tail — the split is the *steadiest* of the three, p90/median
1.01× against the GPU's 1.10×. Consistent with the GPU being the contended
engine rather than the slow one.

## What I think this settles

Put beside the earlier timing sweep — the GPU faster in the middle on 8 of the
12 operators both engines accept, and twice as spread at p99 — the case for
per-operator routing by speed is now weak on this silicon:

- crossings are **not** cheap: 1.67× on the one graph that could be measured
- the faster engine is usually the GPU, which is the one you cannot occupy
  freely because the display composites on it
- so the quantity to minimise is the tail and the time the GPU is held, and
  **neither is minimised by moving operators between engines**

"Send everything the NPU will take, in one partition" comes out ahead of any
split I can construct here. Which makes widening what the NPU takes the useful
work, and on this phone that is constant filters
([NxPU#176](https://github.com/m96-chan/NxPU/pull/176)) rather than a router.

I would not close this on those grounds — one graph on one handset — but I
would stop treating the router as the obvious next thing.

## Two findings on the way

- **`matvec_q8` is refused by both engines**, for different reasons, and had
  never been measured because the operator matrix excludes it for lowering to
  four operators. GPU: `CAST: Not supported Cast case. Input type: INT8 and
  output type: FLOAT32`. NPU: `ANEURALNETWORKS_BAD_DATA at line 1131 while
  adding operation`. Both are about what this compiler emits; filed separately.
- **`executed` is wrong in multi-delegate mode.** Reported on
  [DroidRunner#159](https://github.com/m96-chan/DroidRunner/issues/159).
